Document/0x06c-Reverse-Engineering-and-Tampering.md

```diff
@@ -164,6 +164,109 @@ The following command is listing shared libraries:
+
+We will be analyzing the native function identified in *viewDidLoad()* function in the previous section. The function is at offset *0x1000080d4*. The return value of this function used in *setText* function call for the label. This text is used to compare against the user input. Thus, we can be sure that this function will be returning a string or equivalent. 
+
+![Disassembly of the native function](Images/Chapters/0x06c/manual_reversing_ghidra_native_disassembly.png "Disassembly of the native function")
+
+First thing we can see in the disassembly of the function is there is no input to the function. The registers X0-X7 are not read through out the function. Also, there are multiple calls to other functions like *0x100008158*, *0x10000dbf0* etc. 
+
+Instructions to one such function call are below. Branch instruction *bl* is used to call the function *0x100008158*. 
+
+```
+1000080f0 1a 00 00 94     bl         FUN_100008158                
+1000080f4 60 02 00 39     strb       w0,[x19]=>DAT_10000dbf0
+```
+
+The return value from the function (return value in W0), is stored to the address in register X19 (*strb* stores a byte to the address in register). We can see the same pattern for other function calls, the returned value is stored in X19 register and each time the offset is one more than the previous function call. This behavior can be associated with populating each index of a string array at a time. Each return value is been written to an index of this string array. There are 11 such calls, and from the current evidence we can make an intelligent guess that length of the 'hidden flag' is 11. Towards the end of the disassembly, the function returns with the address to this string array. 
+
+```
+100008148 e0 03 13 aa     mov        x0=>DAT_10000dbf0,x19
+```
+
+To determine the value of the 'hidden flag' we need to know the return value of each of the subsequent function calls identified above. On analysing the function *0x100006fb4*, we can observe this function is much bigger and more complex than previous one we analysed. For analysing complex functions, function graphs can be very helpful, as it helps in understanding the control flow of the function. In Ghidra function graphs can be obtained by clicking **Display function graph** icon in the sub-menu.
+
+![Function graph from 0x100006fb4](Images/Chapters/0x06c/manual_reversing_ghidra_function_graph.png)
+
+Manually analysing all the native functions completely will be time consuming and might not be the wisest approach. In such a scenario using dynamic analysis approach is recommended. By using the techniques like hooking or simply debugging the application, we can easily determine the returned values. In case dynamic analysis approach fails, then we can fallback and start manually analysing the functions. Dynamic analysis techniques are discussed in [Dynamic Analysis](#dynamic-analysis "Dynamic Analysis") section.  
```

Comment:
I've introduced the idea of a feedback loop between static and dynamic. What do you think? 
```suggestion
Manually analysing all the native functions completely will be time consuming and might not be the wisest approach. In such a scenario using a dynamic analysis approach is highly recommended. For instance, by using the techniques like hooking or simply debugging the application, we can easily determine the returned values. Normally it's a good idea to use a dynamic analysis approach and then fallback to manually analysing the functions in a feedback loop. This way you can benefit from both approaches at the same time while saving time and reducing effort. Dynamic analysis techniques are discussed in "[Dynamic Analysis](#dynamic-analysis "Dynamic Analysis")" section.  
```

Reply:
yep. agreed.